surveyor/observation.go

```diff
@@ -246,5 +300,274 @@ func (o *ServiceObsListener) observationHandler(m *nats.Msg) {
+		}
+		// if not a JSON, ignore
+		if filepath.Ext(fs.Name()) != ".json" {
+			return nil
+		}
+		obs, err := NewServiceObservationFromFile(path, s.opts, s.observationMetrics, s.reconnectCtr)
+		for _, existingObservation := range s.observations {
+			// ignore service if it already exists
+			if obs.opts.ServiceName == existingObservation.opts.ServiceName {
+				return fmt.Errorf("service observation with provided service name already exists: %s", obs.opts.ServiceName)
+			}
+		}
+		if err != nil {
+			return fmt.Errorf("could not create observation from %s: %s", path, err)
+		}
+
+		err = obs.Start()
+		if err != nil {
+			return fmt.Errorf("could not start observation from %s: %s", path, err)
+		}
+
+		s.observations = append(s.observations, obs)
+
+	case event.Has(fsnotify.Remove):
+		// directory removed, delete all observations inside and cancel watching this dir
+		if _, ok := s.observationWatchers[path]; ok {
+			for i := 0; ; i++ {
+				if i > len(s.observations)-1 {
+					break
+				}
+				if strings.HasPrefix(s.observations[i].fromFile, path) {
+					s.observations = removeObservation(s.observations, i)
+					i--
+				}
+			}
+			delete(s.observationWatchers, path)
+			return nil
+		}
+		// if not a directory and not a JSON, ignore
+		if filepath.Ext(path) != ".json" {
+			return nil
+		}
+		for i := 0; ; i++ {
+			if i > len(s.observations)-1 {
+				break
+			}
+			if s.observations[i].fromFile == path {
+				s.observations = removeObservation(s.observations, i)
+				i--
+			}
+		}
+	}
+	return nil
+}
+
+func removeObservation(observations []*ServiceObsListener, i int) []*ServiceObsListener {
+	if i >= len(observations) {
+		return observations
+	}
+	observations[i].Stop()
+	if i < len(observations)-1 {
+		observations = append(observations[:i], observations[i+1:]...)
+	} else {
+		observations = observations[:i]
+	}
+	return observations
+}
+
+type ManageObservationOpts func(*ObservationsManager) error
+
+type ObservationErrHandler func(string, error)
+
+type ObservationsManager struct {
+	surveyor           *Surveyor
+	addObservation     chan ObservationConfig
+	deleteObseravation chan string
+	errHandler         ObservationErrHandler
+}
+
+// ObservationsError is a configuration option for [ManageObservations].
+// It sets error handler invoked when adding/deleting observations fails.
+func ObservationsError(handler ObservationErrHandler) ManageObservationOpts {
+	return func(om *ObservationsManager) error {
+		om.errHandler = handler
+		return nil
+	}
+}
+
+// ManageObservastions creates an ObservationManager, allowing for adding/deleting service observations to the surveyor.
+//
+// ManageObservationOpts can be supplied to configure ObservationsManager.
+func (s *Surveyor) ManageObservastions(opts ...ManageObservationOpts) (*ObservationsManager, error) {
+	obsManager := &ObservationsManager{
+		surveyor:           s,
+		addObservation:     make(chan ObservationConfig, 100),
```

Comment:
same here - how do I update an observation?  say I have new creds?  should we do different channels for add/update?  or change the add to an `upsert`?

Reply:
Yeah, I'll add update, that makes sense. I would say we offer a different channel for that, and update by name.